### ISSUE TITLE
New API ItemPaneManager

### ISSUE THREAD

**Issue by windingwind** [zotero/zotero#3380]
An example call:
```js
let registeredID = Zotero.ItemPaneManager.registerView({
    id: "my-view", // ID of the view. Must be unique for each plugin. Will be namespaced with the plugin ID.
    pluginID: "make-it-red@zotero.org", // Plugin ID
    label: "My View", // Label to display
    mode: ["*"], // View mode(s) for which the view should be used. "*" means all modes.
    onReady: ({ type, mode, item, target }) => {
        Zotero.log("View ready!");
        const doc = target.ownerDocument;
        const elem = doc.createElement("div");
        elem.textContent = `Hello World! This is a custom view in ${mode} for ${item.getField("title")}`;
        target.appendChild(elem);
    },
    onActive: ({ type, mode, item, target }) => {
        Zotero.log("View activated!");
    },
    onInactive: ({ type, mode, item, target }) => {
        Zotero.log("View deactivated!");
    },
    onDestroy: ({ type, mode, item, target }) => {
        Zotero.log("View destroyed!");
    },
});
```

**Comment by dstillman** [comment]
- I don't think "View" is right. `registerPane`? I think they can be panes within the item pane, whether they're tabs or sections. Or `registerSubpane`, but that seems a bit verbose. `registerSection` could be forward-looking, but a bit too tied to a particular presentation.
- → `onActivated`, `onDeactivated`, `onDestroyed`
- Deactivated would be switching away from the tab (or collapsing a section) and destroyed would be unloading the plugin?

**Comment by windingwind** [comment]
> Deactivated would be switching away from the tab (or collapsing a section)

Yes

> and destroyed would be unloading the plugin?

Well, I think it's also closing the reader tab, and whenever the <tabpanel> element is removed from its parentNode.

If we decide to go, I will need to slightly change the way we create item panes in reader tabs:

1. When the opened attachment does not have parent item but custom panes are registered, we still create tab panels, instead of `No parent item`;
2. When we create a tab panel in the library or reader, I will use a MutationObserver to trigger the hooks of custom panes;
3. tab panel elements will use attributes to store the necessary information to call the custom pane hooks.

**Comment by dstillman** [comment]
> Well, I think it's also closing the reader tab

Hmm, is it? Do we not have one context pane that gets reused between tabs?

For `onDeactivated`, I suspect most plugins wouldn't have to do anything, because we won't actually pass item details if the tab isn't visible, right? Though, now that I mention it, what actually gets called when the user switches between items? Is that `onReady`? If so, that might be sort of confusing as a name.

**Comment by dstillman** [comment]
Yeah, so I guess generally, let's define exactly what each of these events means and when they'd be called.

**Comment by windingwind** [comment]
Here is my understanding:

ready: When the pane is created and the first rendering happens here. This can happen when the reader is initialized.
activated: When the pane changes from invisible to visible, that is, when the reader tab is selected && the pane is selected.
deactivated: When the pane changes from visible to invisible, opposite to activated.
destroyed: When the pane is removed, triggered by plugin disable or reader tab close.

The `onDeactivated` event can be used to control pane-related background tasks, like periodically checking the syncing status of notes and displaying them (just an example). If the pane is not visible, then these tasks can be paused. Otherwise, there can be tens of background tasks when there are tens of reader tabs. But generally, plugins only need `onReady` to create a static view in the item pane.

**Comment by windingwind** [comment]
> Do we not have one context pane that gets reused between tabs?

I think tabs have their own item panes. We can optimize that if we want, but if we are going to abandon them soon, maybe just leave it as it is and create this API for plugins.

**Comment by dstillman** [comment]
> ready: When the pane is created and the first rendering happens here. This can happen when the reader is initialized.

I don't think we want to encourage plugins to do any work before the pane is actually made visible. I think if a plugin has to do setup, it can do it on `activated` if it hasn't yet done it. So I would be inclined to remove `ready`.

**Comment by windingwind** [comment]
That's OK, but we will notice plugins to manually check if UI things are already there when initializing because `onActivated` can be triggered multiple times.

**Comment by dstillman** [comment]
Or we could save them from having to store a flag by having an `onLoad` that gets called the first time the tab is selected after item pane creation, before the first `onActivated`. I just don't think we want to call anything before the user interacts with it.

**Comment by dstillman** [comment]
So then maybe `onLoad`, `onActivated`, `onDeactivated`, `onUnload`? And `onLoad` gets called when first selected in that item pane.

**Comment by windingwind** [comment]
That's clear!

**Comment by windingwind** [comment]
I think we will add a new event type `onChange` (or `onUpdate`) when the corresponding item changes and the pane is already loaded. An example is switching between different items in the library.

The difference between `onActivated/onDeactivated` and `onChange` is that the former means the item pane is shown/hidden, and the latter means the item pane should be updated if necessary.
